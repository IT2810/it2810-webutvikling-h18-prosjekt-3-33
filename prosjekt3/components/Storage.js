import React from 'react';
import {
    AsyncStorage,
} from 'react-native';


/* Format of the storage:


{ CONTACTS:   [ { name: "Ulrik", age: 22, address: "Bakklandet"  },
                { name: "Sondre", age: 24, address: "Ila"        },
                { name: "Jørgen", age: 95, address: "Møhlenberg" }, 
                ....
              ],
  

  TASKS:      [ { title: "Meeting", date: "22-12-1995" },
                { title: "Dinner", date: "12-7-2995" },

                ...

              ]

  GOALS:      [ {title: "Stop smoking", duration: 5, start: "22-12-1995"}

              ] 
}





*/