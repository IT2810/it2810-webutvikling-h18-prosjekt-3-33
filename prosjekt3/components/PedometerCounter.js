import React from 'react';
import Expo from 'expo';
import { Pedometer } from 'expo';
import { StyleSheet, Text, View, TouchableOpacity } from 'react-native';
import * as Storage from './Storage';
import Colors from '../constants/Colors';

//Pedometer component used by StepGoal to get step data from core motion (iOS) or
//Google Fit (Android)
export default class PedometerCounter extends React.Component {
  state = {
    isPedometerActive: "checking",
    pastStepCount: 0,
    currentStepCount: 0,
    stepGoal: 0,
  };
  //Subscribe to pedometer count and setting state from props (from StepGoal)
  componentDidMount() {
    this._subscribe();
    this.setState({
      stepGoal: this.props.stepGoal
    })
  }
  //Unsubscribe from pedometer count on unmount
  componentWillUnmount() {
    this._unsubscribe();
  }
  //State changes on new props from StepGoal
  componentDidUpdate(prevProps, prevState){
    if(this.props.stepGoal != prevProps.stepGoal){
      this.setState({stepGoal: this.props.stepGoal})
    }
  }
//Subscribing to Core Motion and Google FIt
  _subscribe = () => {
    this._subscription = Pedometer.watchStepCount(result => {
      this.setState({
        currentStepCount: result.steps
      });
    });

    Pedometer.isAvailableAsync().then(
      result => {
        this.setState({
          isPedometerActive: String(result)
        });
      },
      error => {
        this.setState({
          isPedometerActive: "Could not get isPedometerActive: " + error
        });
      }
    );

    const end = new Date();
    const start = new Date();
    start.setDate(end.getDate() - 1);
    Pedometer.getStepCountAsync(start, end).then(
      result => {
        this.setState({ pastStepCount: result.steps });
      },
      error => {
        this.setState({
          pastStepCount: "Could not get stepCount: " + error
        });
      }
    );
  };

  _unsubscribe = () => {
    this._subscription && this._subscription.remove();
    this._subscription = null;
  };

  render() {
    //Renders a progress bar based on progress / goal relationship
    var totalProgress = this.state.pastStepCount + this.state.currentStepCount;
    var goal = this.state.stepGoal;
    return (
      <View style={styles.container}>
        <View style={styles.textContainer}>
          <Text>
            {totalProgress} out of {goal}
          </Text>
        </View>
        <View style={styles.progressBar}>
          <View style={progressBarStyle(totalProgress, goal)}>
          </View>
        </View>

      </View>
    );
  }
}
//Function for calculating the progressbar colors and width
progressBarStyle = function(progress, goal){
  let progressWidth = 0;
  let backgroundColor = '#000';
  let progressGoalRelation = progress/goal;
  if(progressGoalRelation>=1){
    progressWidth = 1;
    backgroundColor = '#75CDA8';
  }
  if (progressGoalRelation < 1 && progressGoalRelation > 0.4){
    progressWidth = progressGoalRelation;
    backgroundColor = '#EDBF2C';
  }
  if(progressGoalRelation<=0.4){
    progressWidth = progressGoalRelation;
    backgroundColor = '#DB0200';
  }
  return{
    flex: progressWidth,
    backgroundColor: backgroundColor,
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    marginTop: 15,
    alignItems: "center",
    justifyContent: "center"
  },
  progressBar: {
    flex: 1,
    flexDirection: 'row',
    width: 300,
    height: 20,
    backgroundColor: '#eee',
  },
  textContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    width: 300,
  }
});

Expo.registerRootComponent(PedometerCounter);
