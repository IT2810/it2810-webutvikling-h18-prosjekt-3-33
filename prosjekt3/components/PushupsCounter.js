import React from 'react';
import Expo from 'expo';
import { StyleSheet, Text, View, TouchableOpacity, Button } from 'react-native';
import * as Storage from './Storage';
import Colors from '../constants/Colors';

export default class PushupsCounter extends React.Component {
  state = {
    pastPushupsCount: 0,
    currentPushupsCount: 0,
    pushupsGoal: 5
  };
  componentDidMount() {
    this.setState({
      pushupsGoal: this.props.pushupsGoal,
    })
    Storage.getGoals().then(goals =>{
      if(goals.hasOwnProperty('pushupsCount')){
        this.setState({pastPushupsCount: goals.pushupsCount})
      }else{
        goals.pushupsCount = 0;
        Storage.storeGoals(goals);
      }
    });
  }
  componentDidUpdate(prevProps, prevState){
    if(this.props.pushupsGoal != prevProps.pushupsGoal){
      this.setState({pushupsGoal: this.props.pushupsGoal})
    }
    if(this.state != prevState){
      Storage.getGoals().then(goals =>{
        goals.pushupsCount = this.state.pastPushupsCount + this.state.currentPushupsCount;
        Storage.storeGoals(goals);
      });
    }
  }

  decrementCounter(totalProgress){
    this.setState({
      currentPushupsCount: (totalProgress-1 < 0)
        ? this.state.currentPushupsCount
        : this.state.currentPushupsCount-1
    })
  }

  incrementCounter(){
    this.setState({currentPushupsCount: this.state.currentPushupsCount+1})
  }

  progressBarStyle(progress, goal){
    let progressWidth = 0;
    let backgroundColor = '#000';
    let progressGoalRelation = progress/goal;
    if(progressGoalRelation>=1){
      progressWidth = 1;
      backgroundColor = '#75CDA8';
    }
    if (progressGoalRelation < 1 && progressGoalRelation > 0.4){
      progressWidth = progressGoalRelation;
      backgroundColor = '#EDBF2C';
    }
    if(progressGoalRelation<=0.4){
      progressWidth = progressGoalRelation;
      backgroundColor = '#DB0200';
    }
    return{
      flex: progressWidth,
      backgroundColor: backgroundColor,
    }
  }

  render() {
    var totalProgress = this.state.pastPushupsCount + this.state.currentPushupsCount;
    var goal = this.state.pushupsGoal;
    return (
      <View style={styles.container}>
        <View style={styles.textContainer}>
          <View style={styles.minusButtonContainer}>
            <Button
              title="-"
              onPress={() => this.decrementCounter(totalProgress)} />
          </View>
          <Text>
            {totalProgress} out of {goal}
          </Text>
          <View style={styles.buttonContainer}>
            <Button
              title="+"
              onPress={() => this.incrementCounter()}
              style={styles.plusButton}/>
          </View>
        </View>
        <View style={styles.progressBar}>
          <View style={this.progressBarStyle(totalProgress, goal)}>
          </View>
        </View>

      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    marginTop: 15,
    alignItems: "center",
    justifyContent: "center"
  },
  progressBar: {
    flex: 1,
    flexDirection: 'row',
    width: 300,
    height: 20,
    backgroundColor: '#eee',
    marginTop: 20,
  },
  textContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: 300,
  },
  plussButton: {
    backgroundColor: '#2db747',
    fontSize: 25,
  },
  buttonContainer: {
    backgroundColor: '#15c129',
    height: 40,
    width: 40,
  },
  minusButtonContainer: {
    backgroundColor: '#c11537',
    height: 40,
    width: 40,
  }
});
