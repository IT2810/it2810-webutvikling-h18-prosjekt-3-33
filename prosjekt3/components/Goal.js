import React from 'react';
import {
  View, Text, StyleSheet, ScrollView

} from 'react-native';
import Colors from '../constants/Colors.js';
import PedometerCounter from './PedometerCounter';

export default class Goal extends React.Component {
  render() {
    return(
      <ScrollView style={styles.container}>
        <View style={styles.goalContainer}>
          <Text style={styles.goalTitle}>Number of steps</Text>
          <PedometerCounter />
        </View>
        <View style={styles.goalContainer}>
          <Text style={styles.goalTitle}>Hours of studing</Text>
        </View>
        <View style={styles.goalContainer}>
          <Text style={styles.goalTitle}>Number of movies watched</Text>
        </View>
      </ScrollView>
    );
  }
}
const styles = StyleSheet.create({
  goalTitle: {
    fontSize: 16,
    color: 'white',
  },
  goalContainer: {
    backgroundColor: Colors.unfinishedGoalColor,
    height: 150,
    padding: 10,
    marginTop: 20
  },
  container: {
    padding: 20
  },
  emptyContainer: {
    height: 80
  }

});
