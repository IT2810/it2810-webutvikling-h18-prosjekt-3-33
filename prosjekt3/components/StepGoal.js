import React from 'react';
import {
  View, Text, StyleSheet, ScrollView, Button, TouchableOpacity

} from 'react-native';
import Colors from '../constants/Colors.js';
import PedometerCounter from './PedometerCounter';

export default class StepGoal extends React.Component {
  state = {
    stepGoal: 0,
  }
  componentDidMount(){
    this.setState({stepGoal: this.props.stepGoal})
  }
  componentDidUpdate(prevProps, prevState){
    if(this.props.stepGoal != prevProps.stepGoal){
      this.setState({stepGoal: this.props.stepGoal})
    }
  }
  render() {
    return(
      <ScrollView style={styles.container}>
        <View style={styles.goalContainer}>
          <Text style={styles.goalTitle}>Number of steps {this.props.stepGoal} </Text>
          <PedometerCounter stepGoal={this.state.stepGoal} />
        </View>
      </ScrollView>
    );
  }
}
const styles = StyleSheet.create({
  goalTitle: {
    fontSize: 16,
    color: 'white',
  },
  goalContainer: {
    backgroundColor: Colors.unfinishedGoalColor,
    height: 150,
    padding: 10,
    marginTop: 20
  },
  container: {
    padding: 20
  },
  emptyContainer: {
    height: 80
  }

});
