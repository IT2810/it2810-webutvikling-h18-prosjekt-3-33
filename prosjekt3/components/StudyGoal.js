import React from 'react';
import {
  View, Text, StyleSheet, ScrollView, Button, TouchableOpacity

} from 'react-native';
import Colors from '../constants/Colors.js';
import PedometerCounter from './PedometerCounter';
import StudyCounter from './StudyCounter';

export default class StepGoal extends React.Component {
  state = {
    studyGoal: 0,
  }
  componentDidMount(){
    this.setState({studyGoal: this.props.studyGoal})
  }
  componentDidUpdate(prevProps, prevState){
    if(this.props.studyGoal != prevProps.studyGoal){
      this.setState({studyGoal: this.props.studyGoal})
    }
  }
  render() {
    return(
      <ScrollView style={styles.container}>
        <View style={styles.goalContainer}>
          <Text style={styles.goalTitle}>Number of hours studied today </Text>
          <StudyCounter studyGoal={this.props.studyGoal} />
        </View>
      </ScrollView>
    );
  }
}
const styles = StyleSheet.create({
  goalTitle: {
    fontSize: 16,
    color: 'white',
  },
  goalContainer: {
    backgroundColor: Colors.goalComponentOne,
    height: 150,
    padding: 10,
    marginTop: 20
  },
  container: {
    padding: 20
  },
  emptyContainer: {
    height: 80
  }

});
