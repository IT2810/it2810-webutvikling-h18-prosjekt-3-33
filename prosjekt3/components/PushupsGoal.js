import React from 'react';
import {
  View, Text, StyleSheet, ScrollView, Button, TouchableOpacity

} from 'react-native';
import Colors from '../constants/Colors.js';
import PushupsCounter from './PushupsCounter';

export default class PushupsGoal extends React.Component {
  state = {
    pushupsGoal: 0,
  }
  componentDidMount(){
    this.setState({pushupsGoal: this.props.pushupsGoal})
  }
  componentDidUpdate(prevProps, prevState){
    if(this.props.studyGoal != prevProps.studyGoal){
      this.setState({pushupsGoal: this.props.pushupsGoal})
    }
  }
  render() {
    return(
      <ScrollView style={styles.container}>
        <View style={styles.goalContainer}>
          <Text style={styles.goalTitle}>Number of pushups today </Text>
          <PushupsCounter pushupsGoal={this.props.pushupsGoal} />
        </View>
      </ScrollView>
    );
  }
}
const styles = StyleSheet.create({
  goalTitle: {
    fontSize: 16,
    color: 'white',
  },
  goalContainer: {
    backgroundColor: Colors.goalComponentOne,
    height: 150,
    padding: 10,
    marginTop: 20
  },
  container: {
    padding: 20
  },
  emptyContainer: {
    height: 80
  }

});
