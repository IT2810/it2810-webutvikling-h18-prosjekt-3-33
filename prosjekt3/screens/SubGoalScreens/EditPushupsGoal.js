import React from 'react';
import {
	View,
	Text,
	TouchableOpacity,
	Button,
	StyleSheet,
	TextInput
} from 'react-native';
import * as Storage from '../../components/Storage';
import Slider from 'react-native-slider';


export default class EditPushupsGoal extends React.Component {
	state = {
		pushupsGoal: 0,
	}
	static navigationOptions = {
		title: '',
  }
	componentDidMount(){
		this.setState({pushupsGoal: this.props.navigation.getParam('pushupsGoal', 0)})
	}
	componentDidUpdate(prevProps, prevState){
		if(this.state.pushupsGoal != prevState.pushupsGoal){
			Storage.getGoals().then(goals => {
				goals['pushupsGoal'] = Number(this.state.pushupsGoal);
				Storage.storeGoals(goals);
			});
			const { navigation } = this.props;
			navigation.state.params.onLoad({ pushupsGoal: this.state.pushupsGoal });
		}
	}
	changePushupsGoal(newGoal){
		this.setState({pushupsGoal: newGoal})
	}
	render() {
		let currentGoal = this.state.pushupsGoal;
		return(

			<View style={styles.container}>
				<View style={styles.smallContainer}>
					<Text>Set a new goal for daily number of pushups </Text>
          <Slider
          value={this.state.pushupsGoal}
          onValueChange={(value) => this.changePushupsGoal(value)}
          minimumValue={0}
          maximumValue={30}
          step={1}/>
				<Text style={{fontSize: 32}}>{this.state.pushupsGoal}</Text>
				</View>
			</View>

		);
	}
}
const styles = StyleSheet.create({
	container: {
		flex: 1,
		alignItems: 'center',
		justifyContent: 'center',
		backgroundColor: '#e8fcb8',
	},
	smallContainer: {
		width: '80%',
    alignItems: "stretch",
	}
});
