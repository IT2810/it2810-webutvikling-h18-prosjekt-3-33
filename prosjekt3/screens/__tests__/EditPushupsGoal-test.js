import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import EditPushupsGoal from '../SubGoalScreens/EditPushupsGoal';
import * as Storage from '../../components/Storage';

describe("EditGoalScreen ", () => {

  it('renders the EditPushupsGoal', async () => {
    const navigation = {
      getParam: () => 5,
      state: {
        params: {
          onLoad: jest.fn()
        }
      }
    }
    const tree = renderer.create(<EditPushupsGoal navigation={navigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it('sets new data', () => {
    const navigation = {
      getParam: () => 5,
      state: {
        params: {
          onLoad: jest.fn()
        }
      }
    }
    const tree = renderer.create(<EditPushupsGoal navigation={navigation} />).getInstance();
    let newGoal = 10
    tree.changePushupsGoal(newGoal);
    expect(tree.state.pushupsGoal).toEqual(newGoal);
  })



});
