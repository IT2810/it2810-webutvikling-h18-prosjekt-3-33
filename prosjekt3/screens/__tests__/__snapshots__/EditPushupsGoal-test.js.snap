// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditGoalScreen  renders the EditPushupsGoal 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e8fcb8",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "width": "80%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Set a new goal for daily number of pushups 
    </Text>
    <View
      animationType="timing"
      onLayout={[Function]}
      onValueChange={[Function]}
      step={1}
      style={
        Array [
          Object {
            "height": 40,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
      thumbTouchSize={
        Object {
          "height": 40,
          "width": 40,
        }
      }
      value={5}
    >
      <View
        onLayout={[Function]}
        renderToHardwareTextureAndroid={true}
        style={
          Array [
            Object {
              "backgroundColor": "#b3b3b3",
            },
            Object {
              "borderRadius": 2,
              "height": 4,
            },
            undefined,
          ]
        }
      />
      <View
        collapsable={undefined}
        renderToHardwareTextureAndroid={true}
        style={
          Object {
            "backgroundColor": "#3f3f3f",
            "borderRadius": 2,
            "height": 4,
            "opacity": 0,
            "position": "absolute",
            "width": 0,
          }
        }
      />
      <View
        collapsable={undefined}
        onLayout={[Function]}
        renderToHardwareTextureAndroid={true}
        style={
          Object {
            "backgroundColor": "#343434",
            "borderRadius": 10,
            "height": 20,
            "opacity": 0,
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": 0,
              },
              Object {
                "translateY": 0,
              },
            ],
            "width": 20,
          }
        }
      />
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        renderToHardwareTextureAndroid={true}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {},
          ]
        }
      />
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 32,
        }
      }
    >
      5
    </Text>
  </View>
</View>
`;
